src/session-pool.ts

```diff
@@ -542,61 +548,60 @@ export class SessionPool extends EventEmitter implements SessionPoolInterface {
   }
 
   /**
-   * Attempts to create a session of a certain type.
+   * Attempts to create a single session of a certain type.
    *
    * @private
    *
    * @param {string} type The desired type to create.
    * @returns {Promise}
    */
-  async _createSession(type: types): Promise<void> {
-    const session = this.database.session();
-    const labels = this.options.labels!;
-
-    this._inventory.borrowed.add(session);
+  _createSession(type: types): Promise<void> {
+    const kind = type === types.ReadOnly ? 'reads' : 'writes';
+    const options = {[kind]: 1};
 
-    const createSession = async (): Promise<void> => {
-      await session.create({labels});
-
-      if (type === types.ReadWrite) {
-        try {
-          await this._prepareTransaction(session);
-        } catch (e) {
-          type = types.ReadOnly;
-        }
-      }
-    };
-
-    try {
-      await this._requests.add(createSession);
-    } catch (e) {
-      this._inventory.borrowed.delete(session);
-      throw e;
-    }
-
-    session.type = type;
-    session.lastUsed = Date.now();
-
-    this._inventory[type].push(session);
-    this._inventory.borrowed.delete(session);
+    return this._createSessions(options);
   }
 
   /**
-   * Attempts to create a session but emits any errors that occur.
+   * Batch creates sessions and prepares any necessary transactions.
    *
    * @private
    *
-   * @emits SessionPool#available
-   * @emits SessionPool#error
-   * @param {string} type The desired type to create.
+   * @param {object} [options] Config specifying how many sessions to create.
    * @returns {Promise}
    */
-  async _createSessionInBackground(type: types): Promise<void> {
-    try {
-      await this._createSession(type);
-      this.emit('available');
-    } catch (e) {
-      this.emit('error', e);
+  async _createSessions({
+    reads = 0,
+    writes = 0,
+  }: CreateSessionsOptions): Promise<void> {
+    const labels = this.options.labels!;
+
+    let needed = reads + writes;
+    this._pending += needed;
+
+    while (needed > 0) {
+      let sessions: Session[] | null = null;
+
+      try {
+        [sessions] = await this.database.batchCreateSessions({
```

Comment:
I seem to recall that we don't want to request more than 100 sessions at a time. I believe this isn't a hard limit but to do with channel affinity of sessions and there being performance implications when we exceed 100 sessions on a single channel. For this reason we might want to limit the number requested to the minimum of 100 or needed?

Reply:
I dug a bit more and for sure the java mock emulates this https://github.com/googleapis/google-cloud-java/blob/d6182e6072baac276c413654094dec77ae0b5d69/google-cloud-clients/google-cloud-spanner/src/test/java/com/google/cloud/spanner/MockSpannerServiceImpl.java#L616

The backend might already only send you 100 even if you ask for more so we may not need to do this client side.

/cc: @skuruppu

Reply:
I found a comment that clarifies this. We can request as many sessions. Returned will be, at most, `min(requested, 100)`. Consider this optional. Though it might be good to document/call out in a comment.

Reply:
Good to know! I put a comment in that mentions this.